File: app/MMarket_old/c/src/data/LogEntry.c
// Copyright 15-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

#include "data/LogEntry.h"

/* .
# Logs entry
LogEntry: serial
  # 1 if 'this' is a error entry.
  is_error: int
  # Time of message. Its format is 'dd/mm/yyyy(hh:mm:ss)'.
  time: char *
  # Log message.
  msg: char *
*/

/*--*/

struct LogEntry_LogEntry {
  int is_error;
  char *time;
  char *msg;
};

LogEntry *logEntry_new (int is_error, char *time, char *msg) {
  LogEntry *this = MALLOC(LogEntry);
  this->is_error = is_error;
  this->time = time;
  this->msg = msg;
  return this;
}

int logEntry_is_error (LogEntry *this) {
  return this->is_error;
}

char *logEntry_time (LogEntry *this) {
  return this->time;
}

char *logEntry_msg (LogEntry *this) {
  return this->msg;
}

Js *logEntry_to_js (LogEntry *this) {
  // Arr[Js]
  Arr *js = arr_new();
  arr_push(js, js_wi((int)this->is_error));
  arr_push(js, js_ws(this->time));
  arr_push(js, js_ws(this->msg));
  return js_wa(js);
}

LogEntry *logEntry_from_js (Js *js) {
  // Arr[Js]
  Arr *a = js_ra(js);
  Js **p = (Js **)arr_start(a);
  LogEntry *this = MALLOC(LogEntry);
  this->is_error = js_ri(*p++);
  this->time = js_rs(*p++);
  this->msg = js_rs(*p++);
  return this;
}

/*--*/

File: app/MMarket_old/js/src/data/LogEntry.js
// Copyright 08-Mar-2020 ºDeme
// GNU General Public License - V3 <http://www.gnu.org/licenses/>

/**
    Log entry
**/

import DateDm from "../dmjs/DateDm.js";

/**
    @private
    @param {string} m
    @param {number} indent
    @param {number} len
    @return {string}
**/
function format (m, indent, len) {
  if (m.trim() === "") return m;

  const r = [];
  m.split("\n").forEach(l => {
    const subr = [];

    while (l.length > len) {
      let line = l.substring(0, len);
      l = l.substring(len);
      const ix = line.lastIndexOf(" ");
      if (ix !== -1 && line.substring(0, ix).trim() !== "") {
        l = line.substring(ix + 1) + l;
        line = line.substring(0, ix);
      }
      subr.push(line);
    }

    if (l.trim() !== "") subr.push(l);
    subr.forEach(subl => { r.push(subl) });
  });

  const ind = " ".repeat(indent);
  return r.join("\n" + ind);
}

/**
    Log entry
**/
export default class LogEntry {
  /**
      @private
      @param {boolean} isError
      @param {string} time
      @param {string} msg
  **/
  constructor (isError, time, msg) {
    this._isError = isError;
    this._time = time;
    this._msg = msg;
  }

  /**
      @return {boolean}
  **/
  get isError () {
    return this._isError;
  }

  /**
      @return {!DateDm}
  **/
  get date () {
    const ix = this._time.indexOf("(");
    return DateDm.fromIso(this._time.substring(0, ix).trim());
  }

  /**
      @return {string}
  **/
  toString () {
    const t = this._time;
    const indent = t.length + 3;
    const len = LogEntry.lineWidth() - indent;
    const sep = this._isError ? " = " : " - ";
    return t + sep + format(this._msg, indent, len);
  }

  /**
      @param {!Array<?>} a
      @return {!LogEntry}
  **/
  static fromJs (a) {
    return new LogEntry(a[0], a[1], a[2]);
  }

  /**
      Number of caracters per line in function 'toString()'
      @return {number}
  **/
  static lineWidth () {
    return 120;
  }

}
